tests/system/Database/Live/UpsertTest.php

```diff
@@ -0,0 +1,647 @@
+                break;
+
+            case 'SQLite3':
+                $expected = <<<'SQL'
+                    INSERT INTO `db_user` (`country`, `email`, `name`)
+                    VALUES ('Iran','ahmadinejad@world.com','Ahmadinejad')
+                    ON CONFLICT(`email`)
+                    DO UPDATE SET
+                    `country` = `excluded`.`country`,
+                    `name` = `excluded`.`name`
+                    SQL;
+                break;
+
+            case 'SQLSRV':
+                $expected = <<<'SQL'
+                    MERGE INTO "test"."dbo"."db_user"
+                    USING (
+                    VALUES ('Iran','ahmadinejad@world.com','Ahmadinejad')
+                    ) "_upsert" ("country", "email", "name")
+                    ON ("test"."dbo"."db_user"."email" = "_upsert"."email")
+                    WHEN MATCHED THEN UPDATE SET
+                    "country" = "_upsert"."country",
+                    "name" = "_upsert"."name"
+                    WHEN NOT MATCHED THEN INSERT ("country", "email", "name")
+                    VALUES ("_upsert"."country", "_upsert"."email", "_upsert"."name");
+                    SQL;
+                break;
+
+            case 'OCI8':
+                $expected = <<<'SQL'
+                    MERGE INTO "db_user"
+                    USING (
+                    SELECT 'Iran' "country", 'ahmadinejad@world.com' "email", 'Ahmadinejad' "name" FROM DUAL
+                    ) "_upsert"
+                    ON ("db_user"."email" = "_upsert"."email")
+                    WHEN MATCHED THEN UPDATE SET
+                    "country" = "_upsert"."country",
+                    "name" = "_upsert"."name"
+                    WHEN NOT MATCHED THEN INSERT ("country", "email", "name")
+                    VALUES  ("_upsert"."country", "_upsert"."email", "_upsert"."name")
+                    SQL;
+                break;
+
+            default:
+                $expected = false;
+                break;
+        }
+
+        $userData = [
+            'email'   => 'ahmadinejad@world.com',
+            'name'    => 'Ahmadinejad',
+            'country' => 'Iran',
+        ];
+
+        $this->assertSame($expected, $this->db->table('user')
+            ->set($userData)
+            ->getCompiledUpsert());
+    }
+
+    public function testGetCompiledUpsertBatch()
+    {
+        $userData = [
+            [
+                'email'   => 'ahmadinejad@example.com',
+                'name'    => 'Ahmadinejad',
+                'country' => 'Iran',
+            ],
+            [
+                'email'   => 'pedro@example.com',
+                'name'    => 'Pedro',
+                'country' => 'El Salvador',
+            ],
+        ];
+
+        $sql = $this->db->table('user')
+            ->setData($userData)
+            ->onConstraint('email')
+            ->getCompiledUpsert();
+
+        $pos = strpos($sql, 'ahmadinejad@example.com');
+        $this->assertNotFalse($pos);
```

Comment:
Please use `assertStringContainsString()` instead of `strpos()`.
```suggestion
        $this->assertStringContainsString('ahmadinejad@example.com', $sql);
```